File: android/app/src/main/java/ir/ilam/inspection/util/PersianDate.kt
package ir.ilam.inspection.util

import java.util.Calendar
import java.util.TimeZone

/**
 * Jalali (Hijri Shamsi) calendar conversion.
 *
 * The database always stores unix milliseconds; every Jalali value in the app is
 * produced here, in the presentation layer only. The algorithm is the Borkowski
 * leap-year table, which is exact for the whole range the app can ever see.
 */
object PersianDate {

    /** Tehran is the single civil timezone the app works in. */
    val zone: TimeZone = TimeZone.getTimeZone("Asia/Tehran")

    private val BREAKS = intArrayOf(
        -61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181, 1210,
        1635, 2060, 2097, 2192, 2262, 2324, 2394, 2456, 3178
    )

    data class Jalali(val year: Int, val month: Int, val day: Int)

    private data class Cal(val leap: Int, val gy: Int, val march: Int)

    private fun jalaliCal(jy: Int): Cal {
        val gy = jy + 621
        var leapJ = -14
        var jp = BREAKS[0]
        require(jy in jp until BREAKS[BREAKS.size - 1]) { "jalali year out of range: $jy" }
        var jump = 0
        for (i in 1 until BREAKS.size) {
            val jm = BREAKS[i]
            jump = jm - jp
            if (jy < jm) break
            leapJ += (jump / 33) * 8 + (jump % 33) / 4
            jp = jm
        }
        var n = jy - jp
        leapJ += (n / 33) * 8 + ((n % 33) + 3) / 4
        if (jump % 33 == 4 && jump - n == 4) leapJ += 1

        val leapG = gy / 4 - ((gy / 100 + 1) * 3) / 4 - 150
        val march = 20 + leapJ - leapG

        if (jump - n < 6) n = n - jump + ((jump + 4) / 33) * 33
        var leap = (((n + 1) % 33) - 1) % 4
        if (leap == -1) leap = 4
        return Cal(leap, gy, march)
    }

    private fun gregorianToDayNumber(gy: Int, gm: Int, gd: Int): Int {
        var d = ((gy + (gm - 8) / 6 + 100100) * 1461) / 4 +
            (153 * ((gm + 9) % 12) + 2) / 5 + gd - 34840408
        d -= ((((gy + 100100 + (gm - 8) / 6) / 100) * 3) / 4) - 752
        return d
    }

    private fun dayNumberToGregorian(dayNumber: Int): Triple<Int, Int, Int> {
        var j = 4 * dayNumber + 139361631
        j += (((4 * dayNumber + 183187720) / 146097) * 3) / 4 * 4 - 3908
        val i = ((j % 1461) / 4) * 5 + 308
        val gd = ((i % 153) / 5) + 1
        val gm = ((i / 153) % 12) + 1
        val gy = j / 1461 - 100100 + (8 - gm) / 6
        return Triple(gy, gm, gd)
    }

    private fun jalaliToDayNumber(jy: Int, jm: Int, jd: Int): Int {
        val r = jalaliCal(jy)
        return gregorianToDayNumber(r.gy, 3, r.march) + (jm - 1) * 31 -
            (jm / 7) * (jm - 7) + jd - 1
    }

    private fun dayNumberToJalali(dayNumber: Int): Jalali {
        val (gy, _, _) = dayNumberToGregorian(dayNumber)
        var jy = gy - 621
        val r = jalaliCal(jy)
        val firstDay = gregorianToDayNumber(r.gy, 3, r.march)
        var k = dayNumber - firstDay
        if (k >= 0) {
            if (k <= 185) return Jalali(jy, 1 + k / 31, (k % 31) + 1)
            k -= 186
        } else {
            jy -= 1
            k += 179
            if (r.leap == 1) k += 1
        }
        return Jalali(jy, 7 + k / 30, (k % 30) + 1)
    }

    fun fromGregorian(gy: Int, gm: Int, gd: Int): Jalali =
        dayNumberToJalali(gregorianToDayNumber(gy, gm, gd))

    fun toGregorian(jy: Int, jm: Int, jd: Int): Triple<Int, Int, Int> =
        dayNumberToGregorian(jalaliToDayNumber(jy, jm, jd))

    fun isLeapYear(jy: Int): Boolean = jalaliCal(jy).leap == 0

    fun monthLength(jy: Int, jm: Int): Int = when {
        jm <= 6 -> 31
        jm <= 11 -> 30
        isLeapYear(jy) -> 30
        else -> 29
    }

    fun of(epochMillis: Long): Jalali {
        val c = Calendar.getInstance(zone)
        c.timeInMillis = epochMillis
        return fromGregorian(c.get(Calendar.YEAR), c.get(Calendar.MONTH) + 1, c.get(Calendar.DAY_OF_MONTH))
    }

    fun toEpochMillis(jy: Int, jm: Int, jd: Int): Long {
        val (gy, gm, gd) = toGregorian(jy, jm, jd)
        val c = Calendar.getInstance(zone)
        c.clear()
        c.set(gy, gm - 1, gd, 0, 0, 0)
        return c.timeInMillis
    }

    fun today(): Jalali = of(System.currentTimeMillis())

    /** `۱۴۰۴/۰۶/۱۴` — the display form used everywhere in the UI. */
    fun format(epochMillis: Long): String = format(of(epochMillis))

    fun format(j: Jalali): String =
        PersianNumbers.toPersian("%04d/%02d/%02d".format(j.year, j.month, j.day))

    /** `۱۴۰۴/۰۶/۱۴ - ۰۹:۳۵` for photo stamps and media captions. */
    fun formatWithTime(epochMillis: Long): String {
        val c = Calendar.getInstance(zone)
        c.timeInMillis = epochMillis
        val time = "%02d:%02d".format(c.get(Calendar.HOUR_OF_DAY), c.get(Calendar.MINUTE))
        return format(epochMillis) + " - " + PersianNumbers.toPersian(time)
    }

    /** `050614` in latin digits — the date segment of a tracking code. */
    fun trackingStamp(epochMillis: Long): String {
        val j = of(epochMillis)
        return "%02d%02d%02d".format(j.year % 100, j.month, j.day)
    }

    /** Whole days between two instants, counted on the civil calendar. */
    fun daysBetween(fromMillis: Long, toMillis: Long): Int {
        val a = startOfDay(fromMillis)
        val b = startOfDay(toMillis)
        return ((b - a) / 86_400_000L).toInt()
    }

    fun startOfDay(epochMillis: Long): Long {
        val c = Calendar.getInstance(zone)
        c.timeInMillis = epochMillis
        c.set(Calendar.HOUR_OF_DAY, 0)
        c.set(Calendar.MINUTE, 0)
        c.set(Calendar.SECOND, 0)
        c.set(Calendar.MILLISECOND, 0)
        return c.timeInMillis
    }

    fun endOfDay(epochMillis: Long): Long = startOfDay(epochMillis) + 86_400_000L - 1
}

File: windows/CryptoInspection.Archive/Util/PersianDate.cs
using System;

namespace CryptoInspection.Archive.Util
{
    /// <summary>
    /// Jalali conversion, the same Borkowski leap table the phone uses. The
    /// database holds unix milliseconds only; every Jalali value in this app is
    /// produced here, for display and for report headers.
    /// </summary>
    public static class PersianDate
    {
        private static readonly int[] Breaks =
        {
            -61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181, 1210,
            1635, 2060, 2097, 2192, 2262, 2324, 2394, 2456, 3178
        };

        private static readonly DateTime Epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

        /// <summary>
        /// Tehran, fixed at +03:30. Iran dropped daylight saving in 2022 and the
        /// archive only ever sees data recorded after that, so a fixed offset is
        /// exact here and avoids the invalid local midnights the old DST rules
        /// produced (the switch happened at 24:00).
        /// </summary>
        public static TimeSpan Offset
        {
            get { return TimeSpan.FromMinutes(210); }
        }

        public struct Jalali
        {
            public int Year;
            public int Month;
            public int Day;

            public Jalali(int year, int month, int day)
            {
                Year = year;
                Month = month;
                Day = day;
            }
        }

        private static void Calculate(int jy, out int leap, out int gy, out int march)
        {
            gy = jy + 621;
            int leapJ = -14;
            int jp = Breaks[0];
            if (jy < jp || jy >= Breaks[Breaks.Length - 1])
            {
                throw new ArgumentOutOfRangeException("jy");
            }

            int jump = 0;
            for (int i = 1; i < Breaks.Length; i++)
            {
                int jm = Breaks[i];
                jump = jm - jp;
                if (jy < jm)
                {
                    break;
                }

                leapJ += (jump / 33) * 8 + (jump % 33) / 4;
                jp = jm;
            }

            int n = jy - jp;
            leapJ += (n / 33) * 8 + ((n % 33) + 3) / 4;
            if (jump % 33 == 4 && jump - n == 4)
            {
                leapJ += 1;
            }

            int leapG = gy / 4 - ((gy / 100 + 1) * 3) / 4 - 150;
            march = 20 + leapJ - leapG;

            if (jump - n < 6)
            {
                n = n - jump + ((jump + 4) / 33) * 33;
            }

            leap = (((n + 1) % 33) - 1) % 4;
            if (leap == -1)
            {
                leap = 4;
            }
        }

        private static int GregorianToDayNumber(int gy, int gm, int gd)
        {
            int d = ((gy + (gm - 8) / 6 + 100100) * 1461) / 4
                + (153 * ((gm + 9) % 12) + 2) / 5 + gd - 34840408;
            d -= ((((gy + 100100 + (gm - 8) / 6) / 100) * 3) / 4) - 752;
            return d;
        }

        private static void DayNumberToGregorian(int dayNumber, out int gy, out int gm, out int gd)
        {
            int j = 4 * dayNumber + 139361631;
            j += (((4 * dayNumber + 183187720) / 146097) * 3) / 4 * 4 - 3908;
            int i = ((j % 1461) / 4) * 5 + 308;
            gd = ((i % 153) / 5) + 1;
            gm = ((i / 153) % 12) + 1;
            gy = j / 1461 - 100100 + (8 - gm) / 6;
        }

        private static int JalaliToDayNumber(int jy, int jm, int jd)
        {
            int leap, gy, march;
            Calculate(jy, out leap, out gy, out march);
            return GregorianToDayNumber(gy, 3, march) + (jm - 1) * 31 - (jm / 7) * (jm - 7) + jd - 1;
        }

        private static Jalali DayNumberToJalali(int dayNumber)
        {
            int gy, gm, gd;
            DayNumberToGregorian(dayNumber, out gy, out gm, out gd);
            int jy = gy - 621;
            int leap, calGy, march;
            Calculate(jy, out leap, out calGy, out march);
            int firstDay = GregorianToDayNumber(calGy, 3, march);
            int k = dayNumber - firstDay;
            if (k >= 0)
            {
                if (k <= 185)
                {
                    return new Jalali(jy, 1 + k / 31, (k % 31) + 1);
                }

                k -= 186;
            }
            else
            {
                jy -= 1;
                k += 179;
                if (leap == 1)
                {
                    k += 1;
                }
            }

            return new Jalali(jy, 7 + k / 30, (k % 30) + 1);
        }

        public static Jalali FromGregorian(int gy, int gm, int gd)
        {
            return DayNumberToJalali(GregorianToDayNumber(gy, gm, gd));
        }

        public static bool IsLeapYear(int jy)
        {
            int leap, gy, march;
            Calculate(jy, out leap, out gy, out march);
            return leap == 0;
        }

        public static int MonthLength(int jy, int jm)
        {
            if (jm <= 6)
            {
                return 31;
            }

            if (jm <= 11)
            {
                return 30;
            }

            return IsLeapYear(jy) ? 30 : 29;
        }

        public static Jalali Of(long epochMillis)
        {
            DateTime local = Epoch.AddMilliseconds(epochMillis) + Offset;
            return FromGregorian(local.Year, local.Month, local.Day);
        }

        public static long ToEpochMillis(int jy, int jm, int jd)
        {
            int dayNumber = JalaliToDayNumber(jy, jm, jd);
            int gy, gm, gd;
            DayNumberToGregorian(dayNumber, out gy, out gm, out gd);
            DateTime utcMidnight = new DateTime(gy, gm, gd, 0, 0, 0, DateTimeKind.Utc) - Offset;
            return (long)(utcMidnight - Epoch).TotalMilliseconds;
        }

        /// <summary>`۱۴۰۴/۰۶/۱۴` — the display form used everywhere.</summary>
        public static string Format(long? epochMillis)
        {
            if (!epochMillis.HasValue)
            {
                return string.Empty;
            }

            Jalali j = Of(epochMillis.Value);
            return PersianNumbers.ToPersian(
                string.Format("{0:0000}/{1:00}/{2:00}", j.Year, j.Month, j.Day));
        }

        public static string FormatWithTime(long epochMillis)
        {
            DateTime local = Epoch.AddMilliseconds(epochMillis) + Offset;
            return Format(epochMillis) + " - " +
                PersianNumbers.ToPersian(string.Format("{0:00}:{1:00}", local.Hour, local.Minute));
        }

        /// <summary>Parses `۱۴۰۴/۰۶/۰۱` or `1404-06-01` into unix milliseconds.</summary>
        public static long? TryParse(string text)
        {
            if (string.IsNullOrWhiteSpace(text))
            {
                return null;
            }

            string[] parts = PersianNumbers.ToLatin(text).Split('/', '-', '.');
            if (parts.Length != 3)
            {
                return null;
            }

            int y, m, d;
            if (!int.TryParse(parts[0].Trim(), out y) ||
                !int.TryParse(parts[1].Trim(), out m) ||
                !int.TryParse(parts[2].Trim(), out d))
            {
                return null;
            }

            if (m < 1 || m > 12 || d < 1 || d > MonthLength(y, m))
            {
                return null;
            }

            return ToEpochMillis(y, m, d);
        }

        public static long EndOfDay(long epochMillis)
        {
            Jalali j = Of(epochMillis);
            return ToEpochMillis(j.Year, j.Month, j.Day) + 86400000L - 1;
        }

        public static long StartOfDay(long epochMillis)
        {
            Jalali j = Of(epochMillis);
            return ToEpochMillis(j.Year, j.Month, j.Day);
        }

        public static long NowMillis()
        {
            return (long)(DateTime.UtcNow - Epoch).TotalMilliseconds;
        }
    }
}
